src/parser/core/modules/EntityStatuses.ts

```diff
@@ -0,0 +1,142 @@
+					if (active === 0) {
+						this.debug('Status not currently active on any targets, starting new window')
+						start = rangeEndpoint.timestamp
+					}
+					active++
+				}
+				if (rangeEndpoint.type === REMOVE) {
+					active--
+					if (active === 0) {
+						uptime += rangeEndpoint.timestamp - (start ?? rangeEndpoint.timestamp)
+						this.debug(`Status ended on all targets.  Total uptime now ${this.parser.formatDuration(uptime, 1)}`)
+					}
+				}
+				return uptime
+			}, 0)
+	}
+
+	getStatusRangesForEvent(statusEvent: BuffTrackingEvent) {
+		this.debug(`Determining active time range for status ${statusEvent.ability.name} - started at: ${this.parser.formatTimestamp(statusEvent.start, 1)} - ended at: ${(statusEvent.end)? this.parser.formatTimestamp(statusEvent.end, 1) : ''}`)
+		return this.splitStatusEventForInvulns(statusEvent).reduce((statusRanges: StatusRangeEndpoint[], statusEvent) => {
+			statusRanges.push({timestamp: statusEvent.start, type: APPLY, stackHistory: statusEvent.stackHistory})
+			statusRanges.push({timestamp: statusEvent.end ?? this.parser.currentTimestamp, type: REMOVE, stackHistory: statusEvent.stackHistory})
+			return statusRanges
+		}, [])
+	}
+
+	splitStatusEventForInvulns(statusEvent: BuffTrackingEvent) {
+		const eventToAdjust = {...statusEvent}
+		const adjustedEvents = [eventToAdjust]
+		if (!statusEvent.end) {
+			this.debug('Unfinished status event detected.  Applying ability duration after last refresh event.')
+			const statusInfo = getDataBy(STATUSES, 'id', statusEvent.ability.guid)
+			if (statusInfo?.duration) {
+				statusEvent.end = statusEvent.lastRefreshed + statusInfo.duration * 1000
+				this.debug(`Updating status event for status ${statusInfo.name}.  Adding ${statusInfo.duration} seconds, effective end time set to ${this.parser.formatTimestamp(statusEvent.end, 1)}`)
+			} else {
+				this.debug(`No matching status duration information found for status ${statusEvent.ability.guid}, setting to end of fight so invuln detection can clip the end to when the target went untargetable`)
+				statusEvent.end = this.parser.fight.end_time
+			}
+		}
+		const target = String(statusEvent.targetID)
+		const invulns = this.invuln.getInvulns(target, statusEvent.start, statusEvent.end, 'invulnerable')
+
+		invulns.forEach((invuln: TODO) => {
+			this.debug(`Target was detected as invulnerable during duration of status.  Invulnerable from ${this.parser.formatTimestamp(invuln.start, 1)} to ${this.parser.formatTimestamp(invuln.end, 1)}`)
+			if (invuln.start < statusEvent.start && invuln.end >= statusEvent.start) {
+				this.debug('Invuln clipped the start of the range - changing beginning of event')
+				eventToAdjust.start = invuln.end
+			} else {
+				this.debug('Invuln clipped the end of the range or split the range - changing end of event')
+				eventToAdjust.end = invuln.start
+				eventToAdjust.stackHistory.splice(-1, 1, {stacks: 0, timestamp: invuln.start, invuln: true})
```

Comment:
This will be modifying the original event's history, as well - `{...obj}` is only a shallow copy. You can do a second shallow copy here with a boring `eventToAdjust.stackHistory = [...eventToAdjust.stackHistory]`, or go nuclear with `const eventToAdjust = _.cloneDeep(statusEvent)` at the top.